// purescript-psa/output/Data.Posix/index.js
// Generated by psc version 0.10.7
"use strict";
var Prelude = require("../Prelude");
var Data_Newtype = require("../Data.Newtype");
var Data_Eq = require("../Data.Eq");
var Data_Ord = require("../Data.Ord");
var Data_Show = require("../Data.Show");
var Data_Semigroup = require("../Data.Semigroup");

// | A user ID (for a process or a file).
var Uid = function (x) {
    return x;
};

// | A process ID.
var Pid = function (x) {
    return x;
};

// | A group ID (for a process or a file).
var Gid = function (x) {
    return x;
};
var showUid = new Data_Show.Show(function (v) {
    return "(Uid " + (Data_Show.show(Data_Show.showInt)(v) + ")");
});
var showPid = new Data_Show.Show(function (v) {
    return "(Pid " + (Data_Show.show(Data_Show.showInt)(v) + ")");
});
var showGid = new Data_Show.Show(function (v) {
    return "(Gid " + (Data_Show.show(Data_Show.showInt)(v) + ")");
});
var ordUid = Data_Ord.ordInt;
var ordPid = Data_Ord.ordInt;
var ordGid = Data_Ord.ordInt;
var newtypeUid = new Data_Newtype.Newtype(function (n) {
    return n;
}, Uid);
var newtypePid = new Data_Newtype.Newtype(function (n) {
    return n;
}, Pid);
var newtypeGid = new Data_Newtype.Newtype(function (n) {
    return n;
}, Gid);
var eqUid = Data_Eq.eqInt;
var eqPid = Data_Eq.eqInt;
var eqGid = Data_Eq.eqInt;
module.exports = {
    Gid: Gid, 
    Pid: Pid, 
    Uid: Uid, 
    newtypePid: newtypePid, 
    eqPid: eqPid, 
    ordPid: ordPid, 
    showPid: showPid, 
    newtypeGid: newtypeGid, 
    eqGid: eqGid, 
    ordGid: ordGid, 
    showGid: showGid, 
    newtypeUid: newtypeUid, 
    eqUid: eqUid, 
    ordUid: ordUid, 
    showUid: showUid
};
