// purescript-psa/output/Node.FS.Stream/index.js
// Generated by psc version 0.10.7
"use strict";
var Prelude = require("../Prelude");
var Data_Maybe = require("../Data.Maybe");
var Data_Function_Uncurried = require("../Data.Function.Uncurried");
var Data_Nullable = require("../Data.Nullable");
var Control_Monad_Eff = require("../Control.Monad.Eff");
var Node_Stream = require("../Node.Stream");
var Node_Path = require("../Node.Path");
var Node_FS = require("../Node.FS");
var Node_FS_Perms = require("../Node.FS.Perms");
var Node_FS_Internal = require("../Node.FS.Internal");
var Data_Semiring = require("../Data.Semiring");
var Control_Semigroupoid = require("../Control.Semigroupoid");
var Data_Function = require("../Data.Function");
var readWrite = (function () {
    var rw = Data_Semiring.add(Node_FS_Perms.semiringPerm)(Node_FS_Perms.read)(Node_FS_Perms.write);
    return Node_FS_Perms.mkPerms(rw)(rw)(rw);
})();
var $$null = Data_Nullable.toNullable(Data_Maybe.Nothing.value);
var nonnull = function ($8) {
    return Data_Nullable.toNullable(Data_Maybe.Just.create($8));
};
var fs = Node_FS_Internal.unsafeRequireFS;

// | Like `fdCreateWriteStream`, but allows you to pass options.
var fdCreateWriteStreamWith = function (opts) {
    return function (fd) {
        return Node_FS_Internal.mkEff(function (v) {
            return fs.createWriteStream($$null, {
                fd: fd, 
                mode: Node_FS_Perms.permsToInt(opts.perms), 
                flags: Node_FS.fileFlagsToNode(opts.flags)
            });
        });
    };
};

// | Create a Readable stream which reads data from the specified file descriptor.
var fdCreateReadStreamWith = function (opts) {
    return function (fd) {
        return Node_FS_Internal.mkEff(function (v) {
            return fs.createReadStream($$null, {
                fd: fd, 
                mode: Node_FS_Perms.permsToInt(opts.perms), 
                flags: Node_FS.fileFlagsToNode(opts.flags), 
                autoClose: opts.autoClose
            });
        });
    };
};
var defaultWriteStreamOptions = {
    flags: Node_FS.W.value, 
    perms: readWrite
};

// | Create a Writable stream which writes data to the specified file

// | descriptor, using the default options.
var fdCreateWriteStream = fdCreateWriteStreamWith(defaultWriteStreamOptions);
var defaultReadStreamOptions = {
    flags: Node_FS.R.value, 
    perms: readWrite, 
    autoClose: true
};

// | Create a Readable stream which reads data to the specified file

// | descriptor, using the default options.
var fdCreateReadStream = fdCreateReadStreamWith(defaultReadStreamOptions);

// | Like `createWriteStream`, but allows you to pass options.
var createWriteStreamWith = function (opts) {
    return function (file) {
        return Node_FS_Internal.mkEff(function (v) {
            return fs.createWriteStream(nonnull(file), {
                mode: Node_FS_Perms.permsToInt(opts.perms), 
                flags: Node_FS.fileFlagsToNode(opts.flags)
            });
        });
    };
};

// | Create a Writable stream which writes data to the specified file, using

// | the default options.
var createWriteStream = createWriteStreamWith(defaultWriteStreamOptions);

// | Create a Readable stream which reads data from the specified file.
var createReadStreamWith = function (opts) {
    return function (file) {
        return Node_FS_Internal.mkEff(function (v) {
            return fs.createReadStream(nonnull(file), {
                mode: Node_FS_Perms.permsToInt(opts.perms), 
                flags: Node_FS.fileFlagsToNode(opts.flags), 
                autoClose: opts.autoClose
            });
        });
    };
};

// | Create a Readable stream which reads data to the specified file, using

// | the default options.
var createReadStream = createReadStreamWith(defaultReadStreamOptions);
module.exports = {
    createReadStream: createReadStream, 
    createReadStreamWith: createReadStreamWith, 
    createWriteStream: createWriteStream, 
    createWriteStreamWith: createWriteStreamWith, 
    defaultReadStreamOptions: defaultReadStreamOptions, 
    defaultWriteStreamOptions: defaultWriteStreamOptions, 
    fdCreateReadStream: fdCreateReadStream, 
    fdCreateReadStreamWith: fdCreateReadStreamWith, 
    fdCreateWriteStream: fdCreateWriteStream, 
    fdCreateWriteStreamWith: fdCreateWriteStreamWith
};
