// purescript-psa/output/Data.Argonaut.Encode.Class/index.js
// Generated by psc version 0.10.7
"use strict";
var Prelude = require("../Prelude");
var Data_Argonaut_Core = require("../Data.Argonaut.Core");
var Data_Either = require("../Data.Either");
var Data_Foldable = require("../Data.Foldable");
var Data_Generic = require("../Data.Generic");
var Data_Int = require("../Data.Int");
var Data_List = require("../Data.List");
var Data_Map = require("../Data.Map");
var Data_Maybe = require("../Data.Maybe");
var Data_String = require("../Data.String");
var Data_StrMap = require("../Data.StrMap");
var Data_Tuple = require("../Data.Tuple");
var Data_Function = require("../Data.Function");
var Data_List_Types = require("../Data.List.Types");
var Control_Semigroupoid = require("../Control.Semigroupoid");
var Control_Category = require("../Control.Category");
var Data_Functor = require("../Data.Functor");
var Data_Unfoldable = require("../Data.Unfoldable");
var Data_Unit = require("../Data.Unit");
var EncodeJson = function (encodeJson) {
    this.encodeJson = encodeJson;
};
var encodeJsonUnit = new EncodeJson(Data_Function["const"](Data_Argonaut_Core.jsonNull));
var encodeJsonJson = new EncodeJson(Control_Category.id(Control_Category.categoryFn));
var encodeJsonJString = new EncodeJson(Data_Argonaut_Core.fromString);
var encodeJsonJNumber = new EncodeJson(Data_Argonaut_Core.fromNumber);
var encodeJsonJBoolean = new EncodeJson(Data_Argonaut_Core.fromBoolean);
var encodeJsonInt = new EncodeJson(function ($33) {
    return Data_Argonaut_Core.fromNumber(Data_Int.toNumber($33));
});
var encodeJson = function (dict) {
    return dict.encodeJson;
};
var encodeJsonArray = function (dictEncodeJson) {
    return new EncodeJson(function (json) {
        return Data_Argonaut_Core.fromArray(Data_Functor.map(Data_Functor.functorArray)(encodeJson(dictEncodeJson))(json));
    });
};
var encodeJsonChar = new EncodeJson(function ($34) {
    return encodeJson(encodeJsonJString)(Data_String.singleton($34));
});
var encodeJsonEither = function (dictEncodeJson) {
    return function (dictEncodeJson1) {
        return new EncodeJson((function () {
            var obj = function (dictEncodeJson2) {
                return function (tag) {
                    return function (x) {
                        return Data_Argonaut_Core.fromObject(Data_StrMap.fromFoldable(Data_List_Types.foldableList)(new Data_List_Types.Cons(new Data_Tuple.Tuple("tag", Data_Argonaut_Core.fromString(tag)), new Data_List_Types.Cons(new Data_Tuple.Tuple("value", encodeJson(dictEncodeJson2)(x)), Data_List_Types.Nil.value))));
                    };
                };
            };
            return Data_Either.either(obj(dictEncodeJson)("Left"))(obj(dictEncodeJson1)("Right"));
        })());
    };
};
var encodeJsonList = function (dictEncodeJson) {
    return new EncodeJson(function ($35) {
        return Data_Argonaut_Core.fromArray(Data_Functor.map(Data_Functor.functorArray)(encodeJson(dictEncodeJson))(Data_List.toUnfoldable(Data_Unfoldable.unfoldableArray)($35)));
    });
};
var encodeJsonMaybe = function (dictEncodeJson) {
    return new EncodeJson(function (v) {
        if (v instanceof Data_Maybe.Nothing) {
            return Data_Argonaut_Core.jsonNull;
        };
        if (v instanceof Data_Maybe.Just) {
            return encodeJson(dictEncodeJson)(v.value0);
        };
        throw new Error("Failed pattern match at Data.Argonaut.Encode.Class line 45, column 3 - line 45, column 33: " + [ v.constructor.name ]);
    });
};
var encodeJsonTuple = function (dictEncodeJson) {
    return function (dictEncodeJson1) {
        return new EncodeJson(function (v) {
            return encodeJson(encodeJsonArray(encodeJsonJson))([ encodeJson(dictEncodeJson)(v.value0), encodeJson(dictEncodeJson1)(v.value1) ]);
        });
    };
};
var encodeMap = function (dictOrd) {
    return function (dictEncodeJson) {
        return function (dictEncodeJson1) {
            return new EncodeJson(function ($36) {
                return encodeJson(encodeJsonList(encodeJsonTuple(dictEncodeJson)(dictEncodeJson1)))(Data_Map.toList($36));
            });
        };
    };
};
var encodeStrMap = function (dictEncodeJson) {
    return new EncodeJson(function ($37) {
        return Data_Argonaut_Core.fromObject(Data_Functor.map(Data_StrMap.functorStrMap)(encodeJson(dictEncodeJson))($37));
    });
};
var gEncodeJson$prime = (function () {
    var addField = function (field) {
        return Data_StrMap.insert(field.recLabel)(gEncodeJson$prime(field.recValue(Data_Unit.unit)));
    };
    return function (v) {
        if (v instanceof Data_Generic.SInt) {
            return Data_Argonaut_Core.fromNumber(Data_Int.toNumber(v.value0));
        };
        if (v instanceof Data_Generic.SString) {
            return Data_Argonaut_Core.fromString(v.value0);
        };
        if (v instanceof Data_Generic.SChar) {
            return Data_Argonaut_Core.fromString(Data_String.singleton(v.value0));
        };
        if (v instanceof Data_Generic.SNumber) {
            return Data_Argonaut_Core.fromNumber(v.value0);
        };
        if (v instanceof Data_Generic.SBoolean) {
            return Data_Argonaut_Core.fromBoolean(v.value0);
        };
        if (v instanceof Data_Generic.SArray) {
            return Data_Argonaut_Core.fromArray(Data_Functor.map(Data_Functor.functorArray)(function ($38) {
                return gEncodeJson$prime((function (v1) {
                    return v1(Data_Unit.unit);
                })($38));
            })(v.value0));
        };
        if (v instanceof Data_Generic.SUnit) {
            return Data_Argonaut_Core.jsonNull;
        };
        if (v instanceof Data_Generic.SProd) {
            return Data_Argonaut_Core.fromObject(Data_StrMap.insert("tag")(encodeJson(encodeJsonJString)(v.value0))(Data_StrMap.singleton("values")(encodeJson(encodeJsonArray(encodeJsonJson))(Data_Functor.map(Data_Functor.functorArray)(function ($39) {
                return gEncodeJson$prime((function (v1) {
                    return v1(Data_Unit.unit);
                })($39));
            })(v.value1)))));
        };
        if (v instanceof Data_Generic.SRecord) {
            return Data_Argonaut_Core.fromObject(Data_Foldable.foldr(Data_Foldable.foldableArray)(addField)(Data_StrMap.empty)(v.value0));
        };
        throw new Error("Failed pattern match at Data.Argonaut.Encode.Class line 26, column 16 - line 39, column 48: " + [ v.constructor.name ]);
    };
})();

// | Encode any `Generic` data structure into `Json`.
var gEncodeJson = function (dictGeneric) {
    return function ($40) {
        return gEncodeJson$prime(Data_Generic.toSpine(dictGeneric)($40));
    };
};
module.exports = {
    EncodeJson: EncodeJson, 
    encodeJson: encodeJson, 
    gEncodeJson: gEncodeJson, 
    "gEncodeJson'": gEncodeJson$prime, 
    encodeJsonMaybe: encodeJsonMaybe, 
    encodeJsonTuple: encodeJsonTuple, 
    encodeJsonEither: encodeJsonEither, 
    encodeJsonUnit: encodeJsonUnit, 
    encodeJsonJBoolean: encodeJsonJBoolean, 
    encodeJsonJNumber: encodeJsonJNumber, 
    encodeJsonInt: encodeJsonInt, 
    encodeJsonJString: encodeJsonJString, 
    encodeJsonJson: encodeJsonJson, 
    encodeJsonChar: encodeJsonChar, 
    encodeJsonArray: encodeJsonArray, 
    encodeJsonList: encodeJsonList, 
    encodeStrMap: encodeStrMap, 
    encodeMap: encodeMap
};
