// purescript-psa/output/Psa.Types/index.js
// Generated by psc version 0.10.7
"use strict";
var Prelude = require("../Prelude");
var Data_Argonaut_Core = require("../Data.Argonaut.Core");
var Data_Argonaut_Decode = require("../Data.Argonaut.Decode");
var Data_Argonaut_Decode_Combinators = require("../Data.Argonaut.Decode.Combinators");
var Data_Argonaut_Encode = require("../Data.Argonaut.Encode");
var Data_StrMap = require("../Data.StrMap");
var Data_StrMap_ST = require("../Data.StrMap.ST");
var Data_Either = require("../Data.Either");
var Data_Maybe = require("../Data.Maybe");
var Data_Set = require("../Data.Set");
var Data_Traversable = require("../Data.Traversable");
var Data_Tuple = require("../Data.Tuple");
var Control_Monad_Eff = require("../Control.Monad.Eff");
var Unsafe_Coerce = require("../Unsafe.Coerce");
var Data_Eq = require("../Data.Eq");
var Data_Ord = require("../Data.Ord");
var Data_Ordering = require("../Data.Ordering");
var Control_Applicative = require("../Control.Applicative");
var Data_Function = require("../Data.Function");
var Data_Functor = require("../Data.Functor");
var Control_Apply = require("../Control.Apply");
var Data_Argonaut_Decode_Class = require("../Data.Argonaut.Decode.Class");
var Control_Bind = require("../Control.Bind");
var Data_Argonaut_Encode_Class = require("../Data.Argonaut.Encode.Class");

// | Relative files paths from the cwd, tagged as either being part of the

// | source files or library files of a project. The `Unknown` variant exists

// | because some psc errors are inter-module and aren't reported with a

// | canonical file.
var Src = (function () {
    function Src(value0) {
        this.value0 = value0;
    };
    Src.create = function (value0) {
        return new Src(value0);
    };
    return Src;
})();

// | Relative files paths from the cwd, tagged as either being part of the

// | source files or library files of a project. The `Unknown` variant exists

// | because some psc errors are inter-module and aren't reported with a

// | canonical file.
var Lib = (function () {
    function Lib(value0) {
        this.value0 = value0;
    };
    Lib.create = function (value0) {
        return new Lib(value0);
    };
    return Lib;
})();

// | Relative files paths from the cwd, tagged as either being part of the

// | source files or library files of a project. The `Unknown` variant exists

// | because some psc errors are inter-module and aren't reported with a

// | canonical file.
var Unknown = (function () {
    function Unknown() {

    };
    Unknown.value = new Unknown();
    return Unknown;
})();
var parsePosition = Data_Maybe.maybe(Control_Applicative.pure(Data_Either.applicativeEither)(Data_Maybe.Nothing.value))(function (obj) {
    return Data_Functor.map(Data_Either.functorEither)(Data_Maybe.Just.create)(Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Data_Functor.map(Data_Either.functorEither)(function (v) {
        return function (v1) {
            return function (v2) {
                return function (v3) {
                    return {
                        startLine: v, 
                        startColumn: v1, 
                        endLine: v2, 
                        endColumn: v3
                    };
                };
            };
        };
    })(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonInt)(obj)("startLine")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonInt)(obj)("startColumn")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonInt)(obj)("endLine")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonInt)(obj)("endColumn")));
});
var maybeProp = function (dictDecodeJson) {
    return function (obj) {
        return function (key) {
            return Data_Maybe.maybe(new Data_Either.Right(Data_Maybe.Nothing.value))(Data_Argonaut_Decode_Class.decodeJson(Data_Argonaut_Decode_Class.decodeJsonMaybe(dictDecodeJson)))(Data_StrMap.lookup(key)(obj));
        };
    };
};
var parseSuggestion = Data_Maybe.maybe(Control_Applicative.pure(Data_Either.applicativeEither)(Data_Maybe.Nothing.value))(function (obj) {
    return Data_Functor.map(Data_Either.functorEither)(Data_Maybe.Just.create)(Control_Apply.apply(Data_Either.applyEither)(Data_Functor.map(Data_Either.functorEither)(function (v) {
        return function (v1) {
            return {
                replacement: v, 
                replaceRange: v1
            };
        };
    })(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonString)(obj)("replacement")))(Control_Bind.bind(Data_Either.bindEither)(maybeProp(Data_Argonaut_Decode_Class.decodeStrMap(Data_Argonaut_Decode_Class.decodeJsonJson))(obj)("replaceRange"))(parsePosition)));
});
var parsePsaError = function (obj) {
    return Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Control_Apply.apply(Data_Either.applyEither)(Data_Functor.map(Data_Either.functorEither)(function (v) {
        return function (v1) {
            return function (v2) {
                return function (v3) {
                    return function (v4) {
                        return function (v5) {
                            return function (v6) {
                                return {
                                    moduleName: v, 
                                    errorCode: v1, 
                                    errorLink: v2, 
                                    message: v3, 
                                    filename: v4, 
                                    position: v5, 
                                    suggestion: v6
                                };
                            };
                        };
                    };
                };
            };
        };
    })(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonMaybe(Data_Argonaut_Decode_Class.decodeJsonString))(obj)("moduleName")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonString)(obj)("errorCode")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonString)(obj)("errorLink")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonString)(obj)("message")))(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonMaybe(Data_Argonaut_Decode_Class.decodeJsonString))(obj)("filename")))(Control_Bind.bind(Data_Either.bindEither)(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonMaybe(Data_Argonaut_Decode_Class.decodeStrMap(Data_Argonaut_Decode_Class.decodeJsonJson)))(obj)("position"))(parsePosition)))(Control_Bind.bind(Data_Either.bindEither)(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeJsonMaybe(Data_Argonaut_Decode_Class.decodeStrMap(Data_Argonaut_Decode_Class.decodeJsonJson)))(obj)("suggestion"))(parseSuggestion));
};
var parsePsaResult = function (obj) {
    return Control_Apply.apply(Data_Either.applyEither)(Data_Functor.map(Data_Either.functorEither)(function (v) {
        return function (v1) {
            return {
                warnings: v, 
                errors: v1
            };
        };
    })(Control_Bind.bind(Data_Either.bindEither)(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeArray(Data_Argonaut_Decode_Class.decodeStrMap(Data_Argonaut_Decode_Class.decodeJsonJson)))(obj)("warnings"))(Data_Traversable.traverse(Data_Traversable.traversableArray)(Data_Either.applicativeEither)(parsePsaError))))(Control_Bind.bind(Data_Either.bindEither)(Data_Argonaut_Decode_Combinators.getField(Data_Argonaut_Decode_Class.decodeArray(Data_Argonaut_Decode_Class.decodeStrMap(Data_Argonaut_Decode_Class.decodeJsonJson)))(obj)("errors"))(Data_Traversable.traverse(Data_Traversable.traversableArray)(Data_Either.applicativeEither)(parsePsaError)));
};
var eqPsaPath = new Data_Eq.Eq(function (v) {
    return function (v1) {
        if (v instanceof Src && v1 instanceof Src) {
            return v.value0 === v1.value0;
        };
        if (v instanceof Lib && v1 instanceof Lib) {
            return v.value0 === v1.value0;
        };
        if (v instanceof Unknown && v1 instanceof Unknown) {
            return true;
        };
        return false;
    };
});
var ordPsaPath = new Data_Ord.Ord(function () {
    return eqPsaPath;
}, function (v) {
    return function (v1) {
        if (v instanceof Src && v1 instanceof Src) {
            return Data_Ord.compare(Data_Ord.ordString)(v.value0)(v1.value0);
        };
        if (v instanceof Src && v1 instanceof Lib) {
            return Data_Ordering.GT.value;
        };
        if (v instanceof Src && v1 instanceof Unknown) {
            return Data_Ordering.GT.value;
        };
        if (v instanceof Lib && v1 instanceof Src) {
            return Data_Ordering.LT.value;
        };
        if (v instanceof Lib && v1 instanceof Lib) {
            return Data_Ord.compare(Data_Ord.ordString)(v.value0)(v1.value0);
        };
        if (v instanceof Lib && v1 instanceof Unknown) {
            return Data_Ordering.GT.value;
        };
        if (v instanceof Unknown && v1 instanceof Unknown) {
            return Data_Ordering.EQ.value;
        };
        if (v instanceof Unknown) {
            return Data_Ordering.LT.value;
        };
        throw new Error("Failed pattern match at Psa.Types line 56, column 3 - line 56, column 40: " + [ v.constructor.name, v1.constructor.name ]);
    };
});
var encodePosition = Unsafe_Coerce.unsafeCoerce;
var encodeSuggestion = function (suggestion) {
    return Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeStrMap(Data_Argonaut_Encode_Class.encodeJsonJson))(Control_Monad_Eff.runPure(Data_StrMap.runST(function __do() {
        var v = Data_StrMap_ST["new"]();
        Data_StrMap_ST.poke(v)("replacement")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJString)(suggestion.replacement))();
        return Data_StrMap_ST.poke(v)("replaceRange")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJson)(Data_Maybe.maybe(Data_Argonaut_Core.jsonNull)(encodePosition)(suggestion.replaceRange)))();
    })));
};
var encodePsaError = function (error) {
    return Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeStrMap(Data_Argonaut_Encode_Class.encodeJsonJson))(Control_Monad_Eff.runPure(Data_StrMap.runST(function __do() {
        var v = Data_StrMap_ST["new"]();
        Data_StrMap_ST.poke(v)("moduleName")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonMaybe(Data_Argonaut_Encode_Class.encodeJsonJString))(error.moduleName))();
        Data_StrMap_ST.poke(v)("errorCode")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJString)(error.errorCode))();
        Data_StrMap_ST.poke(v)("errorLink")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJString)(error.errorLink))();
        Data_StrMap_ST.poke(v)("message")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJString)(error.message))();
        Data_StrMap_ST.poke(v)("filename")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonMaybe(Data_Argonaut_Encode_Class.encodeJsonJString))(error.filename))();
        Data_StrMap_ST.poke(v)("position")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJson)(Data_Maybe.maybe(Data_Argonaut_Core.jsonNull)(encodePosition)(error.position)))();
        return Data_StrMap_ST.poke(v)("suggestion")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonJson)(Data_Maybe.maybe(Data_Argonaut_Core.jsonNull)(encodeSuggestion)(error.suggestion)))();
    })));
};
var encodePsaResult = function (res) {
    return Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeStrMap(Data_Argonaut_Encode_Class.encodeJsonJson))(Control_Monad_Eff.runPure(Data_StrMap.runST(function __do() {
        var v = Data_StrMap_ST["new"]();
        Data_StrMap_ST.poke(v)("warnings")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonArray(Data_Argonaut_Encode_Class.encodeJsonJson))(Data_Functor.map(Data_Functor.functorArray)(encodePsaError)(res.warnings)))();
        return Data_StrMap_ST.poke(v)("errors")(Data_Argonaut_Encode_Class.encodeJson(Data_Argonaut_Encode_Class.encodeJsonArray(Data_Argonaut_Encode_Class.encodeJsonJson))(Data_Functor.map(Data_Functor.functorArray)(encodePsaError)(res.errors)))();
    })));
};
var compareByLocation = function (err1) {
    return function (err2) {
        var $44 = Data_Ord.compare(ordPsaPath)(err1.path)(err2.path);
        if ($44 instanceof Data_Ordering.EQ) {
            if (err1.position instanceof Data_Maybe.Nothing && err2.position instanceof Data_Maybe.Nothing) {
                return Data_Ordering.EQ.value;
            };
            if (err1.position instanceof Data_Maybe.Nothing) {
                return Data_Ordering.LT.value;
            };
            if (err2.position instanceof Data_Maybe.Nothing) {
                return Data_Ordering.GT.value;
            };
            if (err1.position instanceof Data_Maybe.Just && err2.position instanceof Data_Maybe.Just) {
                return Data_Ord.compare(Data_Tuple.ordTuple(Data_Ord.ordInt)(Data_Ord.ordInt))(new Data_Tuple.Tuple(err1.position.value0.startLine, err1.position.value0.startColumn))(new Data_Tuple.Tuple(err2.position.value0.startLine, err2.position.value0.startColumn));
            };
            throw new Error("Failed pattern match at Psa.Types line 117, column 7 - line 123, column 52: " + [ err1.position.constructor.name, err2.position.constructor.name ]);
        };
        return $44;
    };
};
module.exports = {
    Src: Src, 
    Lib: Lib, 
    Unknown: Unknown, 
    compareByLocation: compareByLocation, 
    encodePsaError: encodePsaError, 
    encodePsaResult: encodePsaResult, 
    parsePsaError: parsePsaError, 
    parsePsaResult: parsePsaResult, 
    eqPsaPath: eqPsaPath, 
    ordPsaPath: ordPsaPath
};
