// purescript-psa/output/Data.Monoid/index.js
// Generated by psc version 0.10.7
"use strict";
var Data_Function = require("../Data.Function");
var Data_Semigroup = require("../Data.Semigroup");
var Data_Unit = require("../Data.Unit");

// | A `Monoid` is a `Semigroup` with a value `mempty`, which is both a

// | left and right unit for the associative operation `<>`:

// |

// | ```text

// | forall x. mempty <> x = x <> mempty = x

// | ```

// |

// | `Monoid`s are commonly used as the result of fold operations, where

// | `<>` is used to combine individual results, and `mempty` gives the result

// | of folding an empty collection of elements.
var Monoid = function (__superclass_Data$dotSemigroup$dotSemigroup_0, mempty) {
    this["__superclass_Data.Semigroup.Semigroup_0"] = __superclass_Data$dotSemigroup$dotSemigroup_0;
    this.mempty = mempty;
};
var monoidUnit = new Monoid(function () {
    return Data_Semigroup.semigroupUnit;
}, Data_Unit.unit);
var monoidString = new Monoid(function () {
    return Data_Semigroup.semigroupString;
}, "");
var monoidArray = new Monoid(function () {
    return Data_Semigroup.semigroupArray;
}, [  ]);

// | A `Monoid` is a `Semigroup` with a value `mempty`, which is both a

// | left and right unit for the associative operation `<>`:

// |

// | ```text

// | forall x. mempty <> x = x <> mempty = x

// | ```

// |

// | `Monoid`s are commonly used as the result of fold operations, where

// | `<>` is used to combine individual results, and `mempty` gives the result

// | of folding an empty collection of elements.
var mempty = function (dict) {
    return dict.mempty;
};
var monoidFn = function (dictMonoid) {
    return new Monoid(function () {
        return Data_Semigroup.semigroupFn(dictMonoid["__superclass_Data.Semigroup.Semigroup_0"]());
    }, Data_Function["const"](mempty(dictMonoid)));
};
module.exports = {
    Monoid: Monoid, 
    mempty: mempty, 
    monoidUnit: monoidUnit, 
    monoidFn: monoidFn, 
    monoidString: monoidString, 
    monoidArray: monoidArray
};
