// purescript-psa/output/Control.Comonad.Traced.Trans/index.js
// Generated by psc version 0.10.7

// | This module defines the cowriter comonad transformer, `TracedT`.
"use strict";
var Prelude = require("../Prelude");
var Control_Comonad = require("../Control.Comonad");
var Control_Comonad_Trans_Class = require("../Control.Comonad.Trans.Class");
var Control_Extend = require("../Control.Extend");
var Data_Monoid = require("../Data.Monoid");
var Data_Newtype = require("../Data.Newtype");
var Data_Functor = require("../Data.Functor");
var Data_Function = require("../Data.Function");
var Data_Semigroup = require("../Data.Semigroup");

// | The cowriter comonad transformer.

// |

// | This comonad transformer extends the context of a value in the base comonad so that the value

// | depends on a monoidal position of type `t`.

// |

// | The `ComonadTraced` type class describes the operations supported by this comonad.
var TracedT = function (x) {
    return x;
};

// | Unwrap a value in the `TracedT` comonad.
var runTracedT = function (v) {
    return v;
};
var newtypeTracedT = new Data_Newtype.Newtype(function (n) {
    return n;
}, TracedT);
var functorTracedT = function (dictFunctor) {
    return new Data_Functor.Functor(function (f) {
        return function (v) {
            return Data_Functor.map(dictFunctor)(function (g) {
                return function (t) {
                    return f(g(t));
                };
            })(v);
        };
    });
};
var extendTracedT = function (dictExtend) {
    return function (dictSemigroup) {
        return new Control_Extend.Extend(function () {
            return functorTracedT(dictExtend["__superclass_Data.Functor.Functor_0"]());
        }, function (f) {
            return function (v) {
                return Control_Extend.extend(dictExtend)(function (w$prime) {
                    return function (t) {
                        return f(Data_Functor.map(dictExtend["__superclass_Data.Functor.Functor_0"]())(function (h) {
                            return function (t$prime) {
                                return h(Data_Semigroup.append(dictSemigroup)(t)(t$prime));
                            };
                        })(w$prime));
                    };
                })(v);
            };
        });
    };
};
var comonadTransTracedT = function (dictMonoid) {
    return new Control_Comonad_Trans_Class.ComonadTrans(function (dictComonad) {
        return function (v) {
            return Data_Functor.map((dictComonad["__superclass_Control.Extend.Extend_0"]())["__superclass_Data.Functor.Functor_0"]())(function (f) {
                return f(Data_Monoid.mempty(dictMonoid));
            })(v);
        };
    });
};
var comonadTracedT = function (dictComonad) {
    return function (dictMonoid) {
        return new Control_Comonad.Comonad(function () {
            return extendTracedT(dictComonad["__superclass_Control.Extend.Extend_0"]())(dictMonoid["__superclass_Data.Semigroup.Semigroup_0"]());
        }, function (v) {
            return Control_Comonad.extract(dictComonad)(v)(Data_Monoid.mempty(dictMonoid));
        });
    };
};
module.exports = {
    TracedT: TracedT, 
    runTracedT: runTracedT, 
    newtypeTracedT: newtypeTracedT, 
    functorTracedT: functorTracedT, 
    extendTracedT: extendTracedT, 
    comonadTracedT: comonadTracedT, 
    comonadTransTracedT: comonadTransTracedT
};
