// purescript-psa/output/Ansi.Output/index.js
// Generated by psc version 0.10.7

// | Convenience functions to simplify outputting ANSI escape codes to

// | terminals.
"use strict";
var Prelude = require("../Prelude");
var Ansi_Codes = require("../Ansi.Codes");
var Control_Bind = require("../Control.Bind");

// | Using the given output function, write the given text with the given set

// | of graphics parameters. For example:

// |

// | ```purescript

// | withGraphics Console.log (bold <> underline <> foreground BrightRed) "hello world"

// | ```

// |

// | would print "hello world" to the terminal, bold, underlined, and in bright

// | red, and then reset (so that further logging to the console uses the

// | normal color and style).

// |

// | This function works by first printing the escape code corresponding to the

// | supplied graphics parameters, then printing the text, and then printing a

// | reset escape code.

// |

// | The first argument should take a string, and write it to some output stream

// | such as a terminal. In most cases, you will want to use either

// | `Console.log` or `Console.error`.
var withGraphics = function (dictMonad) {
    return function (write) {
        return function (params) {
            return function (text) {
                return Control_Bind.bind(dictMonad["__superclass_Control.Bind.Bind_1"]())(write(Ansi_Codes.escapeCodeToString(new Ansi_Codes.Graphics(params))))(function () {
                    return Control_Bind.bind(dictMonad["__superclass_Control.Bind.Bind_1"]())(write(text))(function () {
                        return write(Ansi_Codes.escapeCodeToString(new Ansi_Codes.Graphics([ Ansi_Codes.Reset.value ])));
                    });
                });
            };
        };
    };
};
var underline = [ new Ansi_Codes.PMode(Ansi_Codes.Underline.value) ];
var strikethrough = [ new Ansi_Codes.PMode(Ansi_Codes.Strikethrough.value) ];
var italic = [ new Ansi_Codes.PMode(Ansi_Codes.Italic.value) ];
var inverse = [ new Ansi_Codes.PMode(Ansi_Codes.Inverse.value) ];
var foreground = function (c) {
    return [ new Ansi_Codes.PForeground(c) ];
};
var dim = [ new Ansi_Codes.PMode(Ansi_Codes.Dim.value) ];
var bold = [ new Ansi_Codes.PMode(Ansi_Codes.Bold.value) ];
var background = function (c) {
    return [ new Ansi_Codes.PBackground(c) ];
};
module.exports = {
    background: background, 
    bold: bold, 
    dim: dim, 
    foreground: foreground, 
    inverse: inverse, 
    italic: italic, 
    strikethrough: strikethrough, 
    underline: underline, 
    withGraphics: withGraphics
};
