// purescript-psa/output/Node.Platform/index.js
// Generated by psc version 0.10.7

// | This module defines data type for the different platforms supported by

// | Node.js
"use strict";
var Prelude = require("../Prelude");
var Data_Maybe = require("../Data.Maybe");
var Data_Show = require("../Data.Show");
var Data_Eq = require("../Data.Eq");
var Data_Ord = require("../Data.Ord");
var Data_Ordering = require("../Data.Ordering");
var Darwin = (function () {
    function Darwin() {

    };
    Darwin.value = new Darwin();
    return Darwin;
})();
var FreeBSD = (function () {
    function FreeBSD() {

    };
    FreeBSD.value = new FreeBSD();
    return FreeBSD;
})();
var Linux = (function () {
    function Linux() {

    };
    Linux.value = new Linux();
    return Linux;
})();
var SunOS = (function () {
    function SunOS() {

    };
    SunOS.value = new SunOS();
    return SunOS;
})();
var Win32 = (function () {
    function Win32() {

    };
    Win32.value = new Win32();
    return Win32;
})();

// | The String representation for a platform, recognised by Node.js.
var toString = function (v) {
    if (v instanceof Darwin) {
        return "darwin";
    };
    if (v instanceof FreeBSD) {
        return "freebsd";
    };
    if (v instanceof Linux) {
        return "linux";
    };
    if (v instanceof SunOS) {
        return "sunos";
    };
    if (v instanceof Win32) {
        return "win32";
    };
    throw new Error("Failed pattern match at Node.Platform line 17, column 1 - line 18, column 1: " + [ v.constructor.name ]);
};
var showPlatform = new Data_Show.Show(function (v) {
    if (v instanceof Darwin) {
        return "Darwin";
    };
    if (v instanceof FreeBSD) {
        return "FreeBSD";
    };
    if (v instanceof Linux) {
        return "Linux";
    };
    if (v instanceof SunOS) {
        return "SunOS";
    };
    if (v instanceof Win32) {
        return "Win32";
    };
    throw new Error("Failed pattern match at Node.Platform line 32, column 3 - line 33, column 3: " + [ v.constructor.name ]);
});
var fromString = function (v) {
    if (v === "darwin") {
        return new Data_Maybe.Just(Darwin.value);
    };
    if (v === "freebsd") {
        return new Data_Maybe.Just(FreeBSD.value);
    };
    if (v === "linux") {
        return new Data_Maybe.Just(Linux.value);
    };
    if (v === "sunos") {
        return new Data_Maybe.Just(SunOS.value);
    };
    if (v === "win32") {
        return new Data_Maybe.Just(Win32.value);
    };
    return Data_Maybe.Nothing.value;
};
var eqPlatform = new Data_Eq.Eq(function (x) {
    return function (y) {
        if (x instanceof Darwin && y instanceof Darwin) {
            return true;
        };
        if (x instanceof FreeBSD && y instanceof FreeBSD) {
            return true;
        };
        if (x instanceof Linux && y instanceof Linux) {
            return true;
        };
        if (x instanceof SunOS && y instanceof SunOS) {
            return true;
        };
        if (x instanceof Win32 && y instanceof Win32) {
            return true;
        };
        return false;
    };
});
var ordPlatform = new Data_Ord.Ord(function () {
    return eqPlatform;
}, function (x) {
    return function (y) {
        if (x instanceof Darwin && y instanceof Darwin) {
            return Data_Ordering.EQ.value;
        };
        if (x instanceof Darwin) {
            return Data_Ordering.LT.value;
        };
        if (y instanceof Darwin) {
            return Data_Ordering.GT.value;
        };
        if (x instanceof FreeBSD && y instanceof FreeBSD) {
            return Data_Ordering.EQ.value;
        };
        if (x instanceof FreeBSD) {
            return Data_Ordering.LT.value;
        };
        if (y instanceof FreeBSD) {
            return Data_Ordering.GT.value;
        };
        if (x instanceof Linux && y instanceof Linux) {
            return Data_Ordering.EQ.value;
        };
        if (x instanceof Linux) {
            return Data_Ordering.LT.value;
        };
        if (y instanceof Linux) {
            return Data_Ordering.GT.value;
        };
        if (x instanceof SunOS && y instanceof SunOS) {
            return Data_Ordering.EQ.value;
        };
        if (x instanceof SunOS) {
            return Data_Ordering.LT.value;
        };
        if (y instanceof SunOS) {
            return Data_Ordering.GT.value;
        };
        if (x instanceof Win32 && y instanceof Win32) {
            return Data_Ordering.EQ.value;
        };
        throw new Error("Failed pattern match at Node.Platform line 39, column 1 - line 39, column 36: " + [ x.constructor.name, y.constructor.name ]);
    };
});
module.exports = {
    Darwin: Darwin, 
    FreeBSD: FreeBSD, 
    Linux: Linux, 
    SunOS: SunOS, 
    Win32: Win32, 
    fromString: fromString, 
    toString: toString, 
    showPlatform: showPlatform, 
    eqPlatform: eqPlatform, 
    ordPlatform: ordPlatform
};
